### PR TITLE
Add FDS how-to guides

### PR DESCRIPTION
**Pull request by adam-narozniak**
Add Flower Dataset How-To guides
* TensorFlow,
* PyTorch,
* Numpy.

### CODE REVIEW

File: datasets/doc/source/how-to.rst [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+Learn how to transform the HuggingFace dataset to the framework of your choice.
+
+.. toctree::
+   :maxdepth: 1
+
+   how-to-use-with-tf
+   how-to-use-with-torch
+   how-to-use-with-numpy
+
+Feel free to check the original HuggingFace `documentation <https://huggingface.co/docs/datasets/index>`_ if you didn't find the things you were looking for.
```
Comment (danieljanes) [inline datasets/doc/source/how-to.rst:14]: This file can be removed. In framework/baselines, we link to individual how-to guides directly from `index.rst`

File: datasets/doc/source/how-to-use-with-tf.rst [hunk 2]
```diff
@@ -0,0 +1,68 @@
+Use with TensorFlow
```
Comment (danieljanes) [inline datasets/doc/source/how-to-use-with-tf.rst:1]: In the framework docs, we spell out the full framework name also in the filename (`how-to-use-with-tensorflow.rst`)

File: datasets/doc/source/how-to-use-with-torch.rst [hunk 3]
```diff
@@ -0,0 +1,51 @@
+Use with PyTorch
```
Comment (danieljanes) [inline datasets/doc/source/how-to-use-with-torch.rst:1]: In the framework docs, we spell out the full framework name also in the filename (`how-to-use-with-pytorch.rst`)

File: datasets/doc/source/how-to-use-with-numpy.rst [hunk 4]
```diff
@@ -0,0 +1,60 @@
+Use with Numpy
+===================
+
+Let's integrate ``flwr-datasets`` with Numpy.
```
Comment (danieljanes) [inline datasets/doc/source/how-to-use-with-numpy.rst:4]: Let's try to use the brand in the same way the framework/library itself does:
```suggestion
Use with NumPy
===================

Let's integrate ``flwr-datasets`` with NumPy.
```

**Review by danieljanes — CHANGES_REQUESTED** [review]

File: datasets/doc/source/how-to.rst (hunk 1 above)
Comment (adam-narozniak) [reply datasets/doc/source/how-to.rst:14]: Yes, but that's required to have a sidebar that can be clicked and toggled.

File: datasets/doc/source/how-to-use-with-numpy.rst (hunk 4 above)
Comment (adam-narozniak) [reply datasets/doc/source/how-to-use-with-numpy.rst:4]: I fixed it also in the other places

File: datasets/doc/source/how-to.rst (hunk 1 above)
Comment (danieljanes) [reply datasets/doc/source/how-to.rst]: The toggled sidebar hides the content by default, which is why we decided not to use it for framework and baselines docs. We'll use the same approach here.

File: datasets/doc/source/how-to-install-flwr-datasets.rst [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+
+
+Install stable release (pip)
+----------------------------
+
+Stable releases are available on `PyPI <https://pypi.org/project/flwr_datasets/>`_::
+
+  python -m pip install flwr_datasets
+
+For Image Datasets (e.g. MNIST, CIFAR10) ``flwr_datasets`` should be installed with the ``image`` extra::
```
Comment (danieljanes) [inline datasets/doc/source/how-to-install-flwr-datasets.rst:17]:
```suggestion
For vision datasets (e.g. MNIST, CIFAR10) ``flwr_datasets`` should be installed with the ``vision`` extra::
```

File: datasets/doc/source/how-to-install-flwr-datasets.rst [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+
+  python -m pip install flwr_datasets[vision]
+
+For Audio Datasets (e.g. Speech Command) ``flwr_datasets`` should be installed with the ``audio`` extra::
```
Comment (danieljanes) [inline datasets/doc/source/how-to-install-flwr-datasets.rst:21]:
```suggestion
For audio datasets (e.g. Speech Command) ``flwr_datasets`` should be installed with the ``audio`` extra::
```

File: datasets/doc/source/how-to-install-flwr-datasets.rst [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+
+Flower Datasets requires `Python 3.8 <https://docs.python.org/3.8/>`_ or above.
+
+
+Install stable release (pip)
+----------------------------
+
+Stable releases are available on `PyPI <https://pypi.org/project/flwr_datasets/>`_::
+
+  python -m pip install flwr_datasets
```
Comment (danieljanes) [inline datasets/doc/source/how-to-install-flwr-datasets.rst:15]: This also needs to be updated in other parts of the documentation:
```suggestion
  python -m pip install flwr-datasets
```

File: datasets/doc/source/how-to-install-flwr-datasets.rst [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+  python -m pip install flwr_datasets[audio]
+
+
+Verify installation
+-------------------
+
+The following command can be used to verify if Flower Datasets was successfully installed. If everything worked, it should print the version of Flower Datasets to the command line::
+
+  python -c "import flwr_datasets;print(flwr_datasets.__version__)"
+  0.0.1
```
Comment (danieljanes) [inline datasets/doc/source/how-to-install-flwr-datasets.rst:32]: Doesn't the `::` cause syntax highlighting for Python code, but we're showing a bash command here? (same above)

File: datasets/doc/source/how-to-use-with-numpy.rst [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+==============
+
+Let's integrate ``flwr-datasets`` with NumPy.
+
+Prepare the desired partitioning::
+
+  from flwr_datasets import FederatedDataset
+  fds = FederatedDataset(dataset="cifar10", partitioners={"train": 10})
+  partition = fds.load_partition(0, "train")
+  centralized_dataset = fds.load_full("test")
```
Comment (danieljanes) [inline datasets/doc/source/how-to-use-with-numpy.rst:11]:
```suggestion
  from flwr_datasets import FederatedDataset

  fds = FederatedDataset(dataset="cifar10", partitioners={"train": 10})
  partition = fds.load_partition(0, "train")
  centralized_dataset = fds.load_full("test")
```

File: datasets/doc/source/how-to-use-with-numpy.rst [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+  The shape of y_train is: (500,), dtype: int64.
+
+Note that the ``X_train`` values are of type ``uint8``. It is not a problem for the TensorFlow model when passing the
+data as input, but it might remind us to normalize the data - global normalization, pre-channel normalization, or simply
+rescale the data to [0, 1] range::
+
+  X_train = (X_train - X_train.mean()) / X_train.std() # Global normalization
+
+
+CNN Keras Model
```
Comment (danieljanes) [inline datasets/doc/source/how-to-use-with-numpy.rst:35]:
```suggestion
CNN Keras model
```

File: datasets/doc/source/how-to-use-with-numpy.rst [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+  ])
+
+  model.compile(optimizer='adam', loss='sparse_categorical_crossentropy',
+              metrics=['accuracy'])
+  model.fit(X_train, y_train, epochs=20, batch_size=64)
+
+You should see about 98% accuracy on the training data at the end of the training.
+
+Note that I used ``"sparse_categorical_crossentropy"``. Make sure to keep it that way if you don't want to one-hot-encode
+the labels.
```
Comment (danieljanes) [inline datasets/doc/source/how-to-use-with-numpy.rst:60]:
```suggestion
Note that we used ``"sparse_categorical_crossentropy"``. Make sure to keep it that way if you don't want to one-hot-encode
the labels.
```

File: datasets/doc/source/how-to-use-with-pytorch.rst [hunk 12]
```diff
@@ -0,0 +1,66 @@
+Use with PyTorch
+================
+Let's integrate ``flwr-datasets`` with PyTorch DataLoaders and keep your PyTorch Transform applied to the data.
+
+Standard setup - download the dataset, choose the partitioning::
+
+  from flwr_datasets import FederatedDataset
+  fds = FederatedDataset(dataset="cifar10", partitioners={"train": 10})
+  partition = fds.load_partition(0, "train")
+  centralized_dataset = fds.load_full("test")
```
Comment (danieljanes) [inline datasets/doc/source/how-to-use-with-pytorch.rst:10]:
```suggestion
  from flwr_datasets import FederatedDataset

  fds = FederatedDataset(dataset="cifar10", partitioners={"train": 10})
  partition = fds.load_partition(0, "train")
  centralized_dataset = fds.load_full("test")
```

File: datasets/doc/source/how-to-use-with-pytorch.rst [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+  partition_len = len(partition)
+  partition_train = partition[:int(0.8 * partition_len)]
+  partition_test = partition[int(0.8 * partition_len):]
+
+And during the training loop, you need to apply one change. With a typical dataloader you get a list returned for each iteration::
+
+  for batch in all_from_pytorch_dataloader:
+    images, labels = batch
+    # Equivalently
+    images, labels = batch[0], batch[1]
```
Comment (danieljanes) [inline datasets/doc/source/how-to-use-with-pytorch.rst:60]:
```suggestion
    # Or alternatively:
    # images, labels = batch[0], batch[1]
```

File: datasets/doc/source/how-to-use-with-pytorch.rst [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+  partition_train = partition_train_test["train"]
+  partition_test = partition_train_test["test"]
+
+Or you can simply calculate the indices yourself::
+
+  partition_len = len(partition)
+  partition_train = partition[:int(0.8 * partition_len)]
+  partition_test = partition[int(0.8 * partition_len):]
+
+And during the training loop, you need to apply one change. With a typical dataloader you get a list returned for each iteration::
```
Comment (danieljanes) [inline datasets/doc/source/how-to-use-with-pytorch.rst:55]:
```suggestion
And during the training loop, you need to apply one change. With a typical dataloader, you get a list returned for each iteration::
```

File: datasets/doc/source/how-to-use-with-tensorflow.rst [hunk 15]
```diff
@@ -0,0 +1,72 @@
...
+TensorFlow Dataset
+------------------
+Work with ``TensorFlow Dataset`` abstraction.
+
+Standard setup::
+
+  from flwr_datasets import FederatedDataset
+  fds = FederatedDataset(dataset="cifar10", partitioners={"train": 10})
+  partition = fds.load_partition(0, "train")
+  centralized_dataset = fds.load_full("test")
```
Comment (danieljanes) [inline datasets/doc/source/how-to-use-with-tensorflow.rst:24]:
```suggestion
  from flwr_datasets import FederatedDataset

  fds = FederatedDataset(dataset="cifar10", partitioners={"train": 10})
  partition = fds.load_partition(0, "train")
  centralized_dataset = fds.load_full("test")
```

File: datasets/doc/source/how-to-use-with-tensorflow.rst [hunk 16]
```diff
@@ -0,0 +1,72 @@
...
+TensorFlow Tensors
+------------------
+Change the data type to TensorFlow Tensors (it's not the TensorFlow dataset).
+
+Standard setup::
+
+  from flwr_datasets import FederatedDataset
+  fds = FederatedDataset(dataset="cifar10", partitioners={"train": 10})
+  partition = fds.load_partition(0, "train")
+  centralized_dataset = fds.load_full("test")
```
Comment (danieljanes) [inline datasets/doc/source/how-to-use-with-tensorflow.rst:42]:
```suggestion
  from flwr_datasets import FederatedDataset

  fds = FederatedDataset(dataset="cifar10", partitioners={"train": 10})
  partition = fds.load_partition(0, "train")
  centralized_dataset = fds.load_full("test")
```

**Review by danieljanes — CHANGES_REQUESTED** [review]
Looks good, just a few minor comments

File: datasets/doc/source/how-to-install-flwr-datasets.rst (hunk 8 above)
Comment (adam-narozniak) [reply datasets/doc/source/how-to-install-flwr-datasets.rst]: Yes, I've changed it now (and divided the bash and the output of bash as separate blocks)

**Pull request merged**